src/java/org/apache/cassandra/service/accord/AccordStateCache.java

```diff
@@ -278,13 +303,36 @@ public Instance(
             this.nodeFactory = nodeFactory;
         }
 
+        public void register(Listener<K, V> l)
+        {
+            if (listeners == null)
+                listeners = new ArrayList<>();
+            listeners.add(l);
+        }
+
+        public void unregister(Listener<K, V> l)
+        {
+            if (listeners == null)
+                throw new AssertionError("No listeners exist");
+            if (!listeners.remove(l))
+                throw new AssertionError("Listener was not registered");
+            if (listeners.isEmpty())
+                listeners = null;
+        }
+
         public Stream<AccordCachingState<K, V>> stream()
         {
             return cache.entrySet().stream()
-                        .filter(e -> keyClass.isAssignableFrom(e.getKey().getClass()))
```

Comment:
we were matching multiple instances due to CFK stuff having the same type key...